traffic_ops/testing/api/v4/cachegroups_test.go

```diff
@@ -32,99 +35,393 @@ import (
+					utils.CkRequest(utils.NoError(), utils.HasStatus(http.StatusOK), utils.ResponseHasLength(1),
+						ValidateExpectedField("ShortName", "pg2")),
+				},
+				"OK when VALID TOPOLOGY parameter": {
+					nil, TOSession, map[string]string{"topology": "mso-topology"}, nil, nil,
+					utils.CkRequest(utils.NoError(), utils.HasStatus(http.StatusOK)),
+				},
+				"OK when VALID TYPE parameter": {
+					nil, TOSession, map[string]string{"type": "ORG_LOC"},
+					nil, nil,
+					utils.CkRequest(utils.NoError(), utils.HasStatus(http.StatusOK), utils.ResponseLengthGreaterOrEqual(1),
+						ValidateExpectedField("TypeName", "ORG_LOC")),
+				},
+				"EMPTY RESPONSE when INVALID ID parameter": {
+					nil, TOSession, map[string]string{"id": "10000"}, nil, nil,
+					utils.CkRequest(utils.NoError(), utils.HasStatus(http.StatusOK), utils.ResponseHasLength(0)),
+				},
+				"EMPTY RESPONSE when INVALID TYPE parameter": {
+					nil, TOSession, map[string]string{"type": "10000"}, nil, nil,
+					utils.CkRequest(utils.NoError(), utils.HasStatus(http.StatusOK), utils.ResponseHasLength(0)),
+				},
+				"FIRST RESULT when LIMIT=1": {
+					nil, TOSession, map[string]string{"orderby": "id", "limit": "1"}, nil, nil,
+					utils.CkRequest(utils.NoError(), utils.HasStatus(http.StatusOK), ValidatePagination("limit")),
+				},
+				"SECOND RESULT when LIMIT=1 OFFSET=1": {
+					nil, TOSession, map[string]string{"orderby": "id", "limit": "1", "offset": "1"}, nil, nil,
+					utils.CkRequest(utils.NoError(), utils.HasStatus(http.StatusOK), ValidatePagination("offset")),
+				},
+				"SECOND RESULT when LIMIT=1 PAGE=2": {
+					nil, TOSession, map[string]string{"orderby": "id", "limit": "1", "page": "2"}, nil, nil,
+					utils.CkRequest(utils.NoError(), utils.HasStatus(http.StatusOK), ValidatePagination("page")),
+				},
+				"BAD REQUEST when INVALID LIMIT parameter": {
+					nil, TOSession, map[string]string{"limit": "-2"}, nil, nil,
+					utils.CkRequest(utils.HasError(), utils.HasStatus(http.StatusBadRequest)),
+				},
+				"BAD REQUEST when INVALID OFFSET parameter": {
+					nil, TOSession, map[string]string{"limit": "1", "offset": "0"}, nil, nil,
+					utils.CkRequest(utils.HasError(), utils.HasStatus(http.StatusBadRequest)),
+				},
+				"BAD REQUEST when INVALID PAGE parameter": {
+					nil, TOSession, map[string]string{"limit": "1", "page": "0"}, nil, nil,
+					utils.CkRequest(utils.HasError(), utils.HasStatus(http.StatusBadRequest)),
+				},
+				"UNAUTHORIZED when NOT LOGGED IN": {
+					nil, NoAuthTOSession, nil, nil, nil,
+					utils.CkRequest(utils.HasError(), utils.HasStatus(http.StatusUnauthorized)),
+				},
+			},
+			"POST": {
+				"UNAUTHORIZED when NOT LOGGED IN": {
+					nil, NoAuthTOSession, nil, nil, nil,
+					utils.CkRequest(utils.HasError(), utils.HasStatus(http.StatusUnauthorized)),
+				},
+			},
+			"PUT": {
+				"OK when VALID request": {
+					GetCacheGroupId(t, "cachegroup1"), TOSession, nil,
+					map[string]interface{}{
```

Comment:
Since these tests are specifically for the client, I think we should be using the structures used by the client - in this case `tc.CacheGroupNullable`

Reply:
The problem is with `tc.CacheGroupNullable` is you can't do inline pointer declarations.

Reply:
Yes, that's a limitation with which many of us were frustrated, and there are some utility functions in [the `github.com/apache/trafficcontrol/lib/go-util` package like `IntPtr`](https://pkg.go.dev/github.com/apache/trafficcontrol/lib/go-util#IntPtr) that help mitigate that.